lib/src/wiggle_abi/req_impl.rs

```diff
@@ -598,16 +598,19 @@ impl FastlyHttpReq for Session {
     ) -> Result<(u32, ResponseHandle, BodyHandle), Error> {
         let handle: PendingRequestHandle = pending_req_handle.into();
 
-        let outcome = match self.async_item_mut(handle.into())?.is_some() {
-            true => {
-                let resp = self
-                    .take_pending_request(pending_req_handle)?
-                    .recv()
-                    .await?;
-                let (resp_handle, resp_body_handle) = self.insert_response(resp);
-                (1, resp_handle, resp_body_handle)
-            }
-            false => (0, INVALID_REQUEST_HANDLE.into(), INVALID_BODY_HANDLE.into()),
+        let outcome = match self.async_item_mut(handle.into())? {
+            Some(item) => match item.is_ready() {
```

Comment:
Thank you for adding the `is_ready()` here. On reflection I'm a bit unsettled that the previous code didn't lead to any test failures. In particular, [`test_poll()`](https://github.com/fastly/Viceroy/blob/main/test-fixtures/src/bin/upstream-async.rs#L72) over here must be getting backend responses fast enough that we aren't able to effectively test that `poll` doesn't block when the response isn't ready. Adding some sleeps to the [test backends](https://github.com/fastly/Viceroy/blob/a38ed29cf69e7559903e0114b19b6ce446cae97c/cli/tests/integration/upstream_async.rs#L12-L25) probably would be enough to expose that behavior; if you're up for it I would feel much better about landing this if we can show that these functions work as expected.